s3-sink-connector/src/main/java/io/aiven/kafka/connect/s3/S3SinkTask.java

```diff
@@ -86,6 +95,20 @@ public void start(final Map<String, String> props) {
         }
     }
 
+    /**
+     * This determines if the file is key based, and possible to change a single file multiple times per flush or if
+     * it's a roll over file which at each flush is reset.
```

Comment:
Can you explain more about this? What is key based grouping, and why does it mutate the file?

Reply:
Hey @gharris1727 first of all thanks for taking a look!

In terms of the roll over and key grouping.

We have documentation on the key grouping here. (but I will explain below what I am doing.)
[Docs](https://github.com/aindriu-aiven/cloud-storage-connectors-for-apache-kafka/tree/main/s3-sink-connector#grouping-by-the-key)


The S3 Sink (along with all the Sinks provided by Aiven) use a "Record Grouper" this record grouper uses the file.name.template to determine if records should be grouped in a 'changelog' or if they should be grouped by 'Key'.

e.g. {{topic}}-{{partition}}-{{start_offset}} is the default and would cause the Record Grouper to group by changelog.
Changelog means records are appended to the same file and on flush, this causes the record files to be rolled over, to use a new end start_offset

So the original file might be 'logs-0-28.ext' and after flush it will be 'logs-0-45.ext' and each event between offset 28 and 44 will be written to the file for partition 0.

As we don't enforce a max number of events per file, or a max file size the flush (this would be new and is something I am looking at in a separate memory improvements PR) works as a delimiter of sorts to roll the files over.

in compact mode the key looks something like
'{{key}}' or '{{key}}-{{topic}}-{{partition}}' and when matching keys appear it will create a new record or if there is one already existing update the existing record.
To handle this, currently the [record grouper](https://github.com/aindriu-aiven/cloud-storage-connectors-for-apache-kafka/blob/939796f3c592befdce6e7cf9db479adaa12ce0d9/commons/src/main/java/io/aiven/kafka/connect/common/grouper/KeyRecordGrouper.java#L70) removes any existing record and adds the latest record to the file. This then gets written on flush.

The multi part upload does not handle a change to the file so the options are to upload every time and immediately close it to complete the upload or wait until the flush and update the record once.

The latter option is I think, better in terms of API costs if it is possible this could update multiple times over a 30-60s period.

Any questions please let me know.